miner/miner.go

```diff
@@ -96,20 +97,39 @@ func (miner *Miner) update() {
 			}
 			switch ev.Data.(type) {
 			case downloader.StartEvent:
-				atomic.StoreInt32(&miner.canStart, 0)
+				// When syncing begins we pause the miner and set a flag to
+				// ensure calls to Start do not start the miner before sync is
+				// finished.
+
+				// We need to lock over setting canStart, checking Mining and
+				// the call to stop, to prevent the race condition where a
+				// prior concurrent call to Start which has passed all its
+				// checks then calls start after stop is called here. Resulting
+				// in the miner starting immediately after this call to stop
+				// whilst canstart is set to false.
+				miner.startStopMutex.Lock()
+				miner.canStart = false
 				if miner.Mining() {
-					miner.Stop()
-					atomic.StoreInt32(&miner.shouldStart, 1)
+					miner.stop()
 					log.Info("Mining aborted due to sync")
 				}
+				miner.startStopMutex.Unlock()
 			case downloader.DoneEvent, downloader.FailedEvent:
-				shouldStart := atomic.LoadInt32(&miner.shouldStart) == 1
-
-				atomic.StoreInt32(&miner.canStart, 1)
-				atomic.StoreInt32(&miner.shouldStart, 0)
-				if shouldStart {
-					miner.Start(miner.coinbase)
+				// When syncing completes, we start the miner if it is expected
+				// to start, we also unset the flag preventing calls to Start
+				// from starting the miner.
+
+				// We need to lock over both the check on shouldStart and the
+				// call to start, to prevent the race condition where a
+				// concurrent call to Stop occurs between the check of
+				// shouldStart and the call to start resulting in the miner
+				// being started after Stop has been called.
+				miner.startStopMutex.Lock()
+				miner.canStart = true
+				if miner.shouldStart {
```

Comment:
The only difference with original code is `shouldStart` is not set to `false`. Although I guess it's fine because the field is never be used

Reply:
Hi @rjl493456442 I reserved `shouldStart` to indicate the state requested by the user, so it is only set in the public methods `Start` and `Stop` and only read in update. It just seemed simpler that way.